### ISSUE TITLE
Why doesn't the loss decrease during training？

### ISSUE THREAD

**Issue by taoumb** [EdoardoBotta/RQ-VAE-Recommender#11]
Can this code be trained correctly?

**Comment by xy-lin77** [comment]
I have run 50,000 iterations, it seems that the loss is always between 1.2-1.4

**Comment by EdoardoBotta** [comment]
Are you using the default hyperparameters for the model size? If yes, those generate a very small model and I do not expect to do well. I am planning to experiment with bigger model size to ensure the model can be trained.

**Comment by TobinShaw** [comment]
@EdoardoBotta @xy-lin77 @taoumb  I also encountered this issue. Have you made any new attempts and got some findings?

**Comment by EdoardoBotta** [comment]
I have updated the model size and dataset with https://github.com/EdoardoBotta/RQ-VAE-Recommender/commit/a3812fff9daf993505dff55d364fc525ed4df144, adding a textual embedding of the title ad additional features. The loss reflected in the progress bar is now a moving average and it steadily decreases during execution of the training loop.

**Comment by Oytpzy** [comment]
I have run 200,000 iterations, it seems that the loss is always between 0.8-1.2

**Comment by Oytpzy** [comment]
import data.processed
import modules.model

train.iterations=200000
train.learning_rate=0.0003
train.weight_decay=0.035
train.batch_size=256
train.vae_input_dim=768
train.vae_hidden_dims=[512, 256, 128]
train.vae_embed_dim=32
train.vae_n_cat_feats=0
train.vae_codebook_size=256
train.wandb_logging=True
train.model_type=%modules.model.ModelType.ENCODER_DECODER
# https://api.wandb.ai/links/botta-edoardo-carnegie-mellon-university/ilggivkz, https://api.wandb.ai/links/botta-edoardo-carnegie-mellon-university/ufe656js
train.pretrained_rqvae_path="trained_models/rqvae_amazon_beauty/checkpoint_high_entropy.pt"
# Best run: https://api.wandb.ai/links/botta-edoardo-carnegie-mellon-university/xb46itxc
# Checkpoint run: https://api.wandb.ai/links/botta-edoardo-carnegie-mellon-university/6m2fkl1y
# train.pretrained_decoder_path="trained_models/transformer_amazon_beauty/checkpoint_199999.pt"
train.save_dir_root="out/decoder/amazon/"
train.dataset_folder="dataset/amazon"
train.dataset=%data.processed.RecDataset.AMAZON
train.force_dataset_process=False
train.full_eval_every=10000
train.partial_eval_every=5000
train.dataset_split="beauty"
train.dropout_p=0.3
train.attn_heads=6
train.attn_embed_dim=384
train.attn_layers=8
train.decoder_embed_dim=128

This is the gin for the model，but the loss is always between 0.8-1.2

**Comment by EdoardoBotta** [comment]
@Oytpzy Can you paste a loss or wandb link?

**Comment by Oytpzy** [comment]
Thank you for your reply!&nbsp;
I was so inspired by Tiger's work that I hope to replicate yours.
Since wandb is not used in my training process, I will send the deleted log to you in the form of attachment, so as to meet the size requirements of your server attachment. Looking forward to your reply!

------------------ 原始邮件 ------------------
发件人:                                                                                                                        "EdoardoBotta/RQ-VAE-Recommender"                                                                                    ***@***.***&gt;;
发送时间:&nbsp;2025年3月12日(星期三) 下午3:57
***@***.***&gt;;
***@***.******@***.***&gt;;
主题:&nbsp;Re: [EdoardoBotta/RQ-VAE-Recommender] Why doesn't the loss decrease during training？ (Issue #11)



   
@Oytpzy Can you paste a loss or wandb link?

—
Reply to this email directly, view it on GitHub, or unsubscribe.
You are receiving this because you were mentioned.Message ID: ***@***.***&gt;
  EdoardoBotta left a comment (EdoardoBotta/RQ-VAE-Recommender#11)
 
@Oytpzy Can you paste a loss or wandb link?

**Comment by EdoardoBotta** [comment]
This looks fine to me, the loss is not expected to go to 0. I would recommend plotting a loss curve or an exponential moving average and you will see that the loss steadily decreases across training.

**Comment by Oytpzy** [comment]
But the metrics are lower than in the paper，is it because the pre-trained RQ-VAE model is inadequate? "trained_models/rqvae_amazon_beauty/checkpoint_high_entropy.pt"‘s Iter is 139999

**Comment by EdoardoBotta** [comment]
The metrics are lower than the paper, correct. This is still an open question I am investigating and have not been able to fully answer yet.

There seems to be a difference in the with the paper in the RQVAE setup. I have tried training it for longer (~500k iterations) but eventually the codebook collapses and too many items get mapped to a unique semantic ID. This is claimed not to happen in the paper.
